### PR TITLE
Fix printing of `InexactError` for `Inf16` arg and similar

### LINKED ISSUE #51087: InexactError shows `Inf16` as `Inf`
This is slightly misleading:
```julia
julia> UInt128(Inf16)
ERROR: InexactError: UInt128(Inf)
Stacktrace:
 [1] UInt128(x::Float16)
   @ Base ./float.jl:888
 [2] top-level scope
   @ REPL[1]:1
```

The 16 is lost is here:
https://github.com/JuliaLang/julia/blob/b4052a5fa5af2c006612c962d6cf26808575c01b/base/errorshow.jl#L177C5-L177C27

### PR DESCRIPTION
**Pull request by lgoettgens**
Resolves https://github.com/JuliaLang/julia/issues/51087
Closes https://github.com/JuliaLang/julia/pull/51163

Use `show` as pointed out in https://github.com/JuliaLang/julia/pull/51163#discussion_r1328682832.

Furthermore add some tests to make sure that everything works as expected.

### CODE REVIEW

**Comment by lgoettgens** [comment]
Pinging @LilithHafner as she reported the issue in https://github.com/JuliaLang/julia/issues/51087.

File: test/errorshow.jl [hunk 1]
```diff
@@ -1090,3 +1090,15 @@ let e = @test_throws MethodError convert(TypeCompareError{Float64,1}, TypeCompar
     @test  occursin("TypeCompareError{Float64,1}", str)
     @test !occursin("TypeCompareError{Float64{},2}", str) # No {...} for types without params
 end
+
+@testset "InexactError for Inf16 should print '16' (#51087)" begin
+    for IntType in [Int8, Int16, Int32, Int64, Int128, UInt8, UInt16, UInt32, UInt64, UInt128]
+        IntStr = string(IntType)
+        for InfVal in [Inf, Inf16, Inf32, Inf64]
```
Comment (LilithHafner) [inline test/errorshow.jl:1097]: Array constructors promote their elements, so this is equivalent to `fill(Inf, 4)`
```suggestion
        for InfVal in Any[Inf, Inf16, Inf32, Inf64]
```

File: test/errorshow.jl [hunk 2]
```diff
@@ -1090,3 +1090,15 @@ let e = @test_throws MethodError convert(TypeCompareError{Float64,1}, TypeCompar
     @test  occursin("TypeCompareError{Float64,1}", str)
     @test !occursin("TypeCompareError{Float64{},2}", str) # No {...} for types without params
 end
+
+@testset "InexactError for Inf16 should print '16' (#51087)" begin
+    for IntType in [Int8, Int16, Int32, Int64, Int128, UInt8, UInt16, UInt32, UInt64, UInt128]
```
Comment (LilithHafner) [inline test/errorshow.jl:1095]:
```suggestion
    @test sprint(showerror, InexactError(:UInt128, UInt128, Inf16)) == "InexactError: UInt128(Inf16)"
    for IntType in [Int8, Int16, Int32, Int64, Int128, UInt8, UInt16, UInt32, UInt64, UInt128]
```
I like to include a simple test in addition to exhaustive tests, so it's more clear what's going on, and to be absolutely sure we're actually fixing and testing the specific issue that was reported.

**Review by LilithHafner — COMMENTED** [review]
Thank you for implementing a fix! This looks great.

The tests need a few changes to make sure they're actually testing what we want them to test, but the implementation looks solid. Thank you.

Sorry nobody saw this until you mentioned me.

**Review by LilithHafner — APPROVED** [review]

**Pull request merged**